File: game-server/src/main/java/net/dodian/uber/game/model/object/GlobalObject.java
package net.dodian.uber.game.model.object;

import net.dodian.uber.game.model.Position;
import net.dodian.uber.game.model.entity.player.Client;
import net.dodian.uber.game.model.entity.player.Player;

import java.util.concurrent.CopyOnWriteArrayList;

public class GlobalObject {

    private static CopyOnWriteArrayList<Object> globalObject = new CopyOnWriteArrayList<Object>();

    public static CopyOnWriteArrayList<Object> getGlobalObject() {
        return globalObject;
    }

    public static void updateNewObject(Object o) {

    }

    public static void updateOldObject(Object o) {

    }

    public static boolean addGlobalObject(final Object o, int time) {
        if ((o.getAttachment() != null && (long) o.getAttachment() - System.currentTimeMillis() > 0) || hasGlobalObject(o))
            return false;
        o.setAttachment(System.currentTimeMillis() + time);
        globalObject.add(o);
        updateNewObject(o);
        return true;
    }

    public static void updateObject(Client c) {
        for (Object o : globalObject) {
            if (c.withinDistance(o)) {
                if ((long) o.getAttachment() - System.currentTimeMillis() > 0) {
                    updateNewObject(o);
                } else {
                    updateOldObject(o);
                    globalObject.remove(o);
                }
            }
        }
    }

    public static boolean hasGlobalObject(Object o) {
        for (Object ob : globalObject) {
            if (ob.x == o.x && ob.y == o.y)
                return true;
        }
        return false;
    }

    public static Object getGlobalObject(int objectX, int objectY) {
        for (Object ob : globalObject) {
            if (ob.x == objectX && ob.y == objectY)
                return ob;
        }
        return null;
    }

}

File: game-server/src/main/kotlin/net/dodian/uber/game/model/object/GlobalObject.kt
package net.dodian.uber.game.model.`object`

import java.util.concurrent.CopyOnWriteArrayList
import net.dodian.uber.game.model.Position
import net.dodian.uber.game.model.entity.player.Client
import net.dodian.uber.game.model.entity.player.PlayerHandler
import net.dodian.uber.game.systems.zone.ZoneUpdateBus
import net.dodian.utilities.zoneUpdateBatchingEnabled

object GlobalObject {
    private val globalObjects = CopyOnWriteArrayList<Object>()

    @JvmStatic
    fun getGlobalObject(): CopyOnWriteArrayList<Object> = globalObjects

    @JvmStatic
    fun updateNewObject(worldObject: Object) {
        if (zoneUpdateBatchingEnabled) {
            ZoneUpdateBus.queueGlobalObjectNew(
                Position(worldObject.x, worldObject.y, worldObject.z),
                worldObject.id,
                worldObject.face,
                worldObject.type,
            )
            return
        }
        for (player in PlayerHandler.players) {
            val client = player as? Client ?: continue
            if (!client.isActive) {
                continue
            }
            if (client.withinDistance(worldObject)) {
                client.ReplaceObject2(
                    Position(worldObject.x, worldObject.y, worldObject.z),
                    worldObject.id,
                    worldObject.face,
                    worldObject.type,
                )
            }
        }
    }

    @JvmStatic
    fun updateOldObject(worldObject: Object) {
        if (zoneUpdateBatchingEnabled) {
            ZoneUpdateBus.queueGlobalObjectOld(
                Position(worldObject.x, worldObject.y, worldObject.z),
                worldObject.oldId,
                worldObject.face,
                worldObject.type,
            )
            return
        }
        for (player in PlayerHandler.players) {
            val client = player as? Client ?: continue
            if (!client.isActive) {
                continue
            }
            if (client.withinDistance(worldObject)) {
                client.ReplaceObject(
                    worldObject.x,
                    worldObject.y,
                    worldObject.oldId,
                    worldObject.face,
                    worldObject.type,
                )
            }
        }
    }

    @JvmStatic
    fun addGlobalObject(worldObject: Object, time: Int): Boolean {
        val expiresAt = worldObject.getAttachment() as? Long
        if ((expiresAt != null && expiresAt - System.currentTimeMillis() > 0L) || hasGlobalObject(worldObject)) {
            return false
        }
        worldObject.setAttachment(System.currentTimeMillis() + time)
        globalObjects += worldObject
        updateNewObject(worldObject)
        return true
    }

    @JvmStatic
    fun updateObject(client: Client) {
        val now = System.currentTimeMillis()
        for (worldObject in globalObjects) {
            if (!client.withinDistance(worldObject)) {
                continue
            }
            val expiresAt = worldObject.getAttachment() as Long
            if (expiresAt - now > 0L) {
                updateNewObject(worldObject)
            } else {
                updateOldObject(worldObject)
                globalObjects.remove(worldObject)
            }
        }
    }

    @JvmStatic
    fun hasGlobalObject(worldObject: Object): Boolean {
        for (existing in globalObjects) {
            if (existing.x == worldObject.x && existing.y == worldObject.y) {
                return true
            }
        }
        return false
    }

    @JvmStatic
    fun getGlobalObject(objectX: Int, objectY: Int): Object? {
        for (existing in globalObjects) {
            if (existing.x == objectX && existing.y == objectY) {
                return existing
            }
        }
        return null
    }
}
